args/src/lib.rs

```diff
@@ -74,80 +80,880 @@ impl FromStr for InterfaceArg {
-    use hardware::pci::bus::Bus;
-    use hardware::pci::device::Device;
-    use hardware::pci::domain::Domain;
-    use hardware::pci::function::Function;
-    use net::interface::InterfaceName;
 
-    use crate::{InterfaceArg, PortArg};
-    use std::str::FromStr;
+use tracing::instrument;
 
-    #[test]
-    fn test_parse_interface() {
-        // interface + port as PCI address
-        let spec = InterfaceArg::from_str("GbEth1.9000=pci@0000:02:01.7").unwrap();
-        assert_eq!(spec.interface.as_ref(), "GbEth1.9000");
-        assert_eq!(
-            spec.port,
-            Some(PortArg::PCI(PciAddress::new(
-                Domain::from(0),
-                Bus::new(2),
-                Device::try_from(1).unwrap(),
-                Function::try_from(7).unwrap()
-            )))
+use bytecheck::CheckBytes;
+use nix::fcntl::{FcntlArg, FdFlag};
+use nix::{fcntl::SealFlag, sys::memfd::MFdFlags};
+
+pub const DEFAULT_DP_UX_PATH: &str = "/var/run/frr/hh/dataplane.sock";
+pub const DEFAULT_DP_UX_PATH_CLI: &str = "/var/run/dataplane/cli.sock";
+pub const DEFAULT_FRR_AGENT_PATH: &str = "/var/run/frr/frr-agent.sock";
+
+/// A type wrapper around [`std::fs::File`] which is reserved to describe linux [memfd] files.
+///
+/// Our main use case for these files is passing ephemeral, launch-time configuration data to
+/// child the dataplane from the dataplane-init process.
+///
+/// # Note
+///
+/// [`MemFile`] is intended for mutation.  Use the [`MemFile::finalize`] method to create a [`FinalizedMemFile`] to pass
+/// to child processes.
+///
+/// [memfd]: https://man7.org/linux/man-pages/man2/memfd_create.2.html
+#[derive(Debug)]
+pub struct MemFile(std::fs::File);
+
+/// A type wrapper around [`MemFile`] for memfd files which are emphatically NOT intended for any kind of data mutation
+/// ever again.
+///
+/// Multiple protections are in place to deny all attempts to mutat the memory contents of these files.
+/// These protections make this type of file suitable for as-safe-as-practical zero-copy deserialization of data
+/// structure serialized by one process and given to a different process.
+///
+/// Protections inude both basic linux DAC read only mode, as well as write, truncation, and extension sealing, as well
+/// as sealing the seals in place to prevent their removal.
+///
+/// # Note
+///
+/// If these files contain secrets (or even if they don't), it is usually best to mark the file as close-on-exec to
+/// further mitigate opportunities for the data to be corrupted / mutatted.
+/// This task, by its nature, can not be done by the parent process (or the child would not get the file descriptor).
+///
+/// As a consequence, this marking step should be taken as soon as the file is received by the child process.
+/// The (unsafe) method [`FinalizedMemFile::from_fd`] takes this action automatically, and is the recommended way to
+/// receive and read the file from child processes.
+pub struct FinalizedMemFile(MemFile);
+
+impl MemFile {
```

Comment:
Can we split Memfile implementation into its own module and file so everything isn't in lib?

Reply:
I would actually be happy to do that, but I'm afraid of a conflict with @Fredi-raspall's work.  Same goes for @qmonnet's request that I break it into two crates: one for args and the other for memfd.  It is a completely reasonable ask but I don't think it is a great idea till after release

https://github.com/githedgehog/dataplane/issues/972